# 1 Importing Libraries

In [6]:
#Import Pandas Numpy and OS
import pandas as pd
import numpy as np
import os

# 2 Importing Orders and Products Data

In [8]:
df_ords = pd.read_csv('/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/1 Original Data/orders.csv')

In [9]:
df_prods = pd.read_csv('/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/1 Original Data/products.csv')

# 3 Dropping Columns

In [11]:
#Dropping the eval_set Column from the Order Data set
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

# 4 Checking for missing Values in days_since_prior_order column

In [13]:
df_ords['days_since_prior_order'].value_counts (dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

# 5 Renaming Columns

In [15]:
#Renaming order_dow column as orders_day_of_week
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [16]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# 5 Changing Data Types

In [18]:
#Changing order_id data type to string
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [19]:
df_ords['order_id'].dtype

dtype('O')

# 6 Transposing Data

In [21]:
#Importing Departments Data
df_dep = pd.read_csv('/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/1 Original Data/departments.csv')

In [22]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [23]:
#Visually Transposed
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [24]:
#Overwrite Transposed Dataframe
df_dep_t = df_dep.T

In [25]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [26]:
#Reset Index Column
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [27]:
#Create New Header
new_header = df_dep_t.iloc[0]

In [28]:
new_header

0    department
Name: department_id, dtype: object

In [29]:
#Create new dataframe starting from row 1
df_dep_t_new = df_dep_t[1:]

In [30]:
#Add New Header
df_dep_t_new.columns = new_header

In [31]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# 7 Data Dictionary

In [33]:
#Creating a Data Dictionary for department_id column
data_dict = df_dep_t_new.to_dict('index')

In [34]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [35]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [36]:
print(data_dict.get('19'))

{'department': 'snacks'}


# 8 Creating a Subset for department_id #19 'snacks'

In [38]:
df_snacks =  df_prods[df_prods['department_id']==19]

In [39]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [40]:
#Saving df_snacks subset as a dataframe
df_snacks =  df_prods[df_prods['department_id']==19]

In [41]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# Exercise 4.4: Data Wrangling

# 2. Find another identifier variable in the df_ords dataframe that doesn't need to be included in your analysis as a numeric variable and change it to a suitable format.

In [44]:
#Use head() function to check variables.
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0  2539329        1    prior             1                   2   
1  2398795        1    prior             2                   3   
2   473747        1    prior             3                   3   
3  2254736        1    prior             4                   4   
4   431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [45]:
#Check data types
df_ords.dtypes

order_id                   object
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

#user_id doesn't need to be numerical because it is a unique identifier for a customer and thus has no numerical value.

In [47]:
#Changing 'user_id' to a string
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [48]:
#Check to make sure the change was successful
df_ords.dtypes

order_id                   object
user_id                    object
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

# 3 Look for a variable in df_ords with an unintuitive name and change its name without overwriting the dataframe.

In [50]:
#Changing from orders_day_of_week back to order_dow
df_ords.rename(columns = {'orders_day_of_week' : 'orders_dow'}, inplace = True)

In [51]:
#Checking to make sure it worked
df_ords.head()

order_id user_id eval_set  order_number  orders_dow  order_hour_of_day  \
0  2539329       1    prior             1           2                  8   
1  2398795       1    prior             2           3                  7   
2   473747       1    prior             3           3                 12   
3  2254736       1    prior             4           4                  7   
4   431534       1    prior             5           4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 4 Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [53]:
#Using the value_counts function to count the order_hour_of_day variable
df_ords['order_hour_of_day'].value_counts (dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

#The busiest hour to customers placing Instacart orders is 10:00 am

# 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [56]:
#Checking data_dictionary
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

4 = produce department

# 6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [59]:
#Creating subset for breakfast items called df_breakfast
df_breakfast =  df_prods[df_prods['department_id']==14]

In [60]:
#Check to make sure it worked
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

# 7. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [62]:
#Creating subset for dinner party items called df_dinner_party
df_dinner_party = df_prods[df_prods['department_id'].isin([5, 7, 12, 20])]

In [63]:
#Checking to make sure it worked
df_dinner_party

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# 8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

7,650 rows

# 9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [67]:
#Create subset showing all orders placed by user_id 1
user_id_1_data = df_ords[df_ords['user_id'] == '1']

In [68]:
#Check data
user_id_1_data

order_id user_id eval_set  order_number  orders_dow  order_hour_of_day  \
0   2539329       1    prior             1           2                  8   
1   2398795       1    prior             2           3                  7   
2    473747       1    prior             3           3                 12   
3   2254736       1    prior             4           4                  7   
4    431534       1    prior             5           4                 15   
5   3367565       1    prior             6           2                  7   
6    550135       1    prior             7           1                  9   
7   3108588       1    prior             8           1                 14   
8   2295261       1    prior             9           1                 16   
9   2550362       1    prior            10           4                  8   
10  1187899       1    train            11           4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# 10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [70]:
#Use describe function to get descriptive stats for user_id_1_data
user_id_1_data.describe()

order_number  orders_dow  order_hour_of_day  days_since_prior_order
count     11.000000   11.000000          11.000000               10.000000
mean       6.000000    2.636364          10.090909               19.000000
std        3.316625    1.286291           3.477198                9.030811
min        1.000000    1.000000           7.000000                0.000000
25%        3.500000    1.500000           7.500000               14.250000
50%        6.000000    3.000000           8.000000               19.500000
75%        8.500000    4.000000          13.000000               26.250000
max       11.000000    4.000000          16.000000               30.000000

1. Customer 1 has ordered from Istacart 11 times
2. On average they order every 19 days
3. They usually order around 10:00 am
4. They always order between 7:00 am - 4:00 pm
5. They always place orders between Monday and Thursday

# 12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [73]:
import os
df_ords.to_csv(os.path.join("/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/2 Prepared Data", "orders_prepared.csv"), index=False)

# 13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [75]:
import os
df_dep_t_new.to_csv(os.path.join("/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/2 Prepared Data", "departments.wrangled.csv"), index=False)

In [123]:
df_ords.shape

(3421083, 7)